File: apps/web/src/lib/mediapipe/pose-sampler.ts
import type { HandFrame } from "@/types/landmarks";

type PoseLandmarks = HandFrame["poseLandmarks"];

export type PoseSampler = {
  sample: (timestamp: number, detect: () => PoseLandmarks) => PoseLandmarks;
  reset: () => void;
};

// Pose moves slowly relative to hands, so detection runs at a reduced rate and
// the last good pose bridges the gap between samples and short detector misses.
export function createPoseSampler({
  sampleMs,
  reuseMs,
}: {
  sampleMs: number;
  reuseMs: number;
}): PoseSampler {
  let pose: PoseLandmarks = [];
  let poseAt = 0;
  let sampledAt = 0;

  return {
    sample(timestamp, detect) {
      if (pose.length === 0 || timestamp - sampledAt >= sampleMs) {
        sampledAt = timestamp;
        const detected = detect();
        if (detected.length > 0) {
          pose = detected;
          poseAt = timestamp;
        }
      }
      if (pose.length > 0 && timestamp - poseAt > reuseMs) pose = [];
      return pose;
    },
    reset() {
      pose = [];
      poseAt = 0;
      sampledAt = 0;
    },
  };
}

File: apps/mobile/HandWave/Sources/Services/PoseSampler.swift
/// Pose moves slowly relative to hands, so detection runs at a reduced rate and
/// the last good pose bridges the gap between samples and short detector misses.
struct PoseSampler {
  private static let sampleIntervalMs = 1_000 / 12
  private static let reuseMs = 500

  private var lastPose: [LandmarkPoint]?
  private var lastPoseAt = 0
  private var lastSampleAt = 0

  mutating func sample(
    at timestampMs: Int,
    detect: () throws -> [LandmarkPoint]?
  ) rethrows -> [LandmarkPoint]? {
    if lastSampleAt == 0 || timestampMs - lastSampleAt >= Self.sampleIntervalMs {
      lastSampleAt = timestampMs
      if let pose = try detect() {
        lastPose = pose
        lastPoseAt = timestampMs
      }
    }
    guard let lastPose, timestampMs - lastPoseAt <= Self.reuseMs else { return nil }
    return lastPose
  }

  mutating func reset() {
    self = PoseSampler()
  }
}
